features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/OpennmsKafkaProducer.java

```diff
@@ -191,7 +199,25 @@ public boolean shouldForwardAlarm(OnmsAlarm alarm) {
         return true;
     }
 
+    private boolean isIncrementalAlarm(String reductionKey, OnmsAlarm alarm) {
+        return outstandingAlarms.containsKey(reductionKey) &&
```

Comment:
Multiple threads can call `updateAlarm` at the same time, so we need to make sure that the usage of the `outstandingAlarms` map is thread safe.

Reply:
It's possible for the call to `equalsExcludingOnFirst` to throw an NPE since the map may change between the call to `containsKey` and `get`.

Reply:
In this case the map may change, but not in such a way that the key no longer exists since I'm not aging entries out. So I don't think it could throw an NPE. However I will change it so that if we decide to age entries out in the future this won't break.

Reply:
Right. We should be deleting alarms from the map then, otherwise the map will just continue to grow.

I would update the `updateAlarm` function to just delete the entry from the map when invoked with a `null` alarm.